packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -2394,20 +2401,23 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin, Resto
           return;
         }
         assert(_currentBottomSheet!._widget == bottomSheet);
-        if (!isPersistent && !removedEntry) {
-          assert(entry != null);
-          entry!.remove();
-          removedEntry = true;
-        }
+        completer.complete();
       },
       onDismissed: () {
         if (_dismissedBottomSheets.contains(bottomSheet)) {
           setState(() {
             _dismissedBottomSheets.remove(bottomSheet);
           });
         }
+        if (_currentBottomSheet == null) {
+          return;
+        }
+        setState(() {
+          _currentBottomSheet = null;
+        });
       },
       onDispose: () {
+        _removeEntryIfNeeded();
```

Comment:
`ShowBottomSheet` multiple times will only call `onDispose`, so this is a good time to remove the entry.